### PR TITLE
Universal [camera stack] refactoring renderer support.

### PR DESCRIPTION
**Pull request by martint-unity**
---
### Purpose of this PR
We are fixing bugs in where it is confusing for our users in using different renderer types and camera stacking.
[FB Case](https://fogbugz.unity3d.com/f/cases/1376601/)
We can not support different renderer types in a stack at the moment. 
This PR adds Editor UI warnings and also makes the run-time not support these cases anymore.

image

---
### Testing status
Tested in Editor to create different overlay cameras and different renderers, Differed, Forward.
Made a test build to make sure it also looks as in editor.

### CODE REVIEW

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 1]
```diff
@@ -92,24 +98,72 @@ void DrawElementCallback(Rect rect, int index, bool isActive, bool isFocused)
 
             (Camera camera, UniversalRenderPipelineSerializedCamera serializedCamera) overlayCamera = m_SerializedCamera[index];
             Camera cam = overlayCamera.camera;
+            var baseAdditionalData = camera.GetUniversalAdditionalCameraData();
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:101]: Can we move this underneath the `if (cam != null)` ?

File: com.unity.render-pipelines.universal/Runtime/UniversalAdditionalCameraData.cs [hunk 2]
```diff
@@ -352,7 +352,7 @@ public CameraRenderType renderType
 
         /// <summary>
         /// Returns the camera stack. Only valid for Base cameras.
-        /// Overlay cameras have no stack and will return null.
+        /// None Base cameras, does not have a stack and will return null.
```
Comment (ellioman) [inline com.unity.render-pipelines.universal/Runtime/UniversalAdditionalCameraData.cs:355]: This sentence feels a bit weird...

**Review by ellioman — APPROVED** [review]
Looks good.

**Review by Nzollner — APPROVED** [review]
LGTM

**Review by kennytann — APPROVED** [review]

File: com.unity.render-pipelines.universal/Runtime/UniversalAdditionalCameraData.cs (hunk 2 above)
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Runtime/UniversalAdditionalCameraData.cs]: fixed

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs (hunk 1 above)
Comment (martint-unity) [reply com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs]: fixed

**Review by ernestasKupciunas — COMMENTED** [review]
Looks good!  Warnings were displayed in the Console. UI warnings too.
Screenshot 2022-02-08 at 14 33 33
I was running URP Rendering Examples project for the regression testing.
In addition, ran this Yamato [job](https://yamato.cds.internal.unity3d.com/jobs/1724-reach-qa-graphics-template-urp/tree/master/.yamato%252FAll.yml%2523trunk_Run_All/11482563/job/pipeline)

- One concern is if we should display the UI Warning only when clicking on the Overlay camera (not sure how it works on Windows machine, since I have only macBook at home)?
ezgif com-gif-maker


Tested using:
2022.2.0a5.838
Revision: trunk ae3a32483326
Built: Sat, 05 Feb 2022 08:23:21 GMT

MacBook Pro (16-inch, 2021), CPU: Apple M1 Max, Memory: 64 GB, macOS: Monterey 12.1

**Review by TheoWong-pixel — APPROVED** [review]

**Comment by martint-unity** [comment]
@ernestasKupciunas Yes, only show the error when selecting the camera was requested by UX.

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs [hunk 3]
```diff
@@ -55,6 +55,27 @@ public static class RenderPass
...
+        /// This setting controls if the camera editor should display the camera stack category.
+        /// If your renderer is not supporting stacking this one should return 0.
+        /// For the UI to show the Camera Stack widget this must support CameraRenderType.Base.
+        /// <see cref="CameraRenderType"/>
+        /// Returns the bitmask of the supported camera render types in the renderer's current state.
+        /// </summary>
+        public virtual int SupportedCameraRenderTypes()
+        {
+            return 0;
+        }
```
Comment (phi-lira) [inline com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:68]: Maybe a better API name could help with confusion but not sure at this point. My confusion I think is that each property in the field means something else:
 - A renderer that says a cameraRenderType.Base is supported means that this renderer supports stacking.
 - A renderer that says a cameraRenderType.Overlay is supported means that this renderer can stack on base.

**Comment by Nzollner** [comment]
If a camera doesn't support stacking or being an overlay then the render type field is disabled.
image

File: com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs [hunk 4]
```diff
@@ -92,24 +98,71 @@ void DrawElementCallback(Rect rect, int index, bool isActive, bool isFocused)
...
             if (cam != null)
             {
-                bool typeError = false;
-                var type = cam.gameObject.GetComponent<UniversalAdditionalCameraData>().renderType;
-                if (type != CameraRenderType.Overlay)
+                bool outputWarning = false;
+
+                // Checking if the Base Camera and the overlay cameras are of the same type.
+                // If not, we report an error.
+                var overlayAdditionalData = cam.gameObject.GetComponent<UniversalAdditionalCameraData>();
```
[the remark this answers is not in the document]
Comment (Nzollner) [reply com.unity.render-pipelines.universal/Editor/Camera/UniversalRenderPipelineCameraEditor.cs:109]: Done in next commit

File: com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs (hunk 3 above)
Comment (Nzollner) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:68]: We have tried to rename the function to:
`SupportedCameraStackingTypes`

Comment (Nzollner) [reply com.unity.render-pipelines.universal/Runtime/ScriptableRenderer.cs:68]: @phi-lira

**Comment by Nzollner** [comment]
Error has been resolved and test is running: 
https://unity-ci.cds.internal.unity3d.com/job/11664642/logs

**Comment by Nzollner** [comment]
It show the incorrect version for yamato test
The correct version is green: https://unity-ci.cds.internal.unity3d.com/job/11664642/dependency-graph

**Review by phi-lira — APPROVED** [review]
LGTM.

**Pull request merged**